### ISSUE TITLE
Running cluster test times out and locks up my computer

### ISSUE THREAD

**Issue by legoktm** [openzim/libzim#147]
I started working on updating the Debian packaging for libzim and while running tests on 3.2.0+, the cluster test would timeout and lock up my computer for about 30 seconds.

I did a git bisect and tracked it down to 655ff7b6bf4ef7154287231739e53d25cabafb60.

```
km@km-pt:~/projects/libzim/build$ ninja test
[0/1] Running all tests.
1/5 cluster                                 TIMEOUT 30.14 s
2/5 dirent                                  OK       0.01 s
3/5 header                                  OK       0.01 s
4/5 uuid                                    OK       1.01 s
5/5 template                                OK       0.01 s

OK:         4
FAIL:       1
SKIP:       0
TIMEOUT:    1


The output from the failed tests:

1/5 cluster                                 TIMEOUT 30.14 s

--- command ---
MAIL='/var/mail/km' USER='km' COMP_WORDBREAKS=' 	
"'><;|&(:' SHLVL='1' HOME='/home/km' OLDPWD='/home/km/projects' HUSHLOGIN='FALSE' DBUS_SESSION_BUS_ADDRESS='unix:path=/run/user/1000/bus' DEBEMAIL='legoktm@member.fsf.org' DEBFULLNAME='Kunal Mehta' LOGNAME='km' JOURNAL_STREAM='9:1566593' _='/usr/bin/ninja' XDG_SESSION_ID='4' TERM='vt220' PATH='/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games' INVOCATION_ID='d14301e573714059bf85e636ac4c84c5' S_COLORS='auto' XDG_RUNTIME_DIR='/run/user/1000' LS_COLORS='' SHELL='/bin/bash' LC_ALL='C.UTF-8' GPG_TTY='/dev/pts/0' PWD='/home/km/projects/libzim/build' /home/km/projects/libzim/build/test/cluster
--- stdout ---
[==========] Running 6 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 5 tests from ClusterTest
[ RUN      ] ClusterTest.create_cluster
[       OK ] ClusterTest.create_cluster (0 ms)
[ RUN      ] ClusterTest.read_write_cluster
[       OK ] ClusterTest.read_write_cluster (0 ms)
[ RUN      ] ClusterTest.read_write_empty
[       OK ] ClusterTest.read_write_empty (0 ms)
[ RUN      ] ClusterTest.read_write_clusterZ
[       OK ] ClusterTest.read_write_clusterZ (1 ms)
[ RUN      ] ClusterTest.read_write_clusterLzma
[       OK ] ClusterTest.read_write_clusterLzma (25 ms)
[----------] 5 tests from ClusterTest (26 ms total)

[----------] 1 test from CluterTest
[ RUN      ] CluterTest.read_write_extended_cluster
-------

Full log written to /home/km/projects/libzim/build/meson-logs/testlog.txt
FAILED: meson-test 
/usr/bin/python3 -u /usr/bin/meson test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.
```
Please let me know if there's any other information I can provide.

**Comment by kelson42** [comment]
@mgautierfr Any clue?

**Comment by legoktm** [comment]
@kelson42 I cherry-picked the fix from ecf8e52d224040acdf108761cb2e20eee209a2df on top of 3.3.0, and the test no longer times out, it just fails:
```
   dh_auto_test -O--buildsystem=meson
	cd obj-x86_64-linux-gnu && LC_ALL=C.UTF-8 MESON_TESTTHREADS=1 ninja test
[0/1] Running all tests.
1/5 cluster                                 FAIL    12.69 s
2/5 dirent                                  OK       0.06 s
3/5 header                                  OK       0.01 s
4/5 uuid                                    OK       1.01 s
5/5 template                                OK       0.01 s

OK:         4
FAIL:       1
SKIP:       0
TIMEOUT:    0


The output from the failed tests:

1/5 cluster                                 FAIL    12.69 s

--- command ---
DEB_HOST_GNU_SYSTEM='linux-gnu' SUDO_GID='1000' DEB_TARGET_GNU_CPU='x86_64' DEB_BUILD_ARCH_BITS='64' MAIL='/var/mail/km' DEB_HOST_ARCH_OS='linux' USER='pbuilder' CXXFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong -Wformat -Werror=format-security' DEB_TARGET_MULTIARCH='x86_64-linux-gnu' DEB_BUILD_GNU_TYPE='x86_64-linux-gnu' OBJCFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong -Wformat -Werror=format-security' DEB_HOST_ARCH_ABI='base' SHLVL='5' DEB_BUILD_ARCH_CPU='amd64' DEB_HOST_ARCH_LIBC='gnu' HOME='/nonexistent' OLDPWD='/' HUSHLOGIN='FALSE' PERL_USE_UNSAFE_INC='1' BUILDRESULTGID='1000' DEB_BUILD_ARCH_ENDIAN='little' LDFLAGS='-Wl,-z,relro' COWDANCER_ILISTFILE='/.ilist' DEB_TARGET_ARCH_BITS='64' DBUS_SESSION_BUS_ADDRESS='unix:path=/run/user/1000/bus' PBUILDER_OPERATION='build' MAKEFLAGS='' DEB_BUILD_GNU_SYSTEM='linux-gnu' DEB_BUILD_ARCH_OS='linux' DEBEMAIL='legoktm@member.fsf.org' DEB_TARGET_GNU_TYPE='x86_64-linux-gnu' DEB_TARGET_ARCH_CPU='amd64' SUDO_UID='1000' PBUILDER_PKGDATADIR='/usr/share/pbuilder' DEBFULLNAME='Kunal Mehta' LOGNAME='pbuilder' JOURNAL_STREAM='9:4716254' _='/usr/bin/unshare' DEB_BUILD_ARCH_LIBC='gnu' PBUILDER_SYSCONFDIR='/etc' DEB_BUILD_ARCH_ABI='base' DEB_TARGET_ARCH_ENDIAN='little' USERNAME='root' XDG_SESSION_ID='4' DEB_HOST_ARCH='amd64' TERM='vt220' DEB_TARGET_GNU_SYSTEM='linux-gnu' DEB_HOST_GNU_CPU='x86_64' PATH='/usr/share/pkg-kde-tools/bin:/usr/sbin:/usr/bin:/sbin:/bin' INVOCATION_ID='4e5a58801b824b32bdc252cc469a27a3' MESON_TESTTHREADS='1' XDG_RUNTIME_DIR='/run/user/1000' DEB_TARGET_ARCH_OS='linux' S_COLORS='auto' MAKELEVEL='1' CFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong -Wformat -Werror=format-security' BUILDRESULTUID='1000' SOURCE_DATE_EPOCH='1527232381' DEB_HOST_MULTIARCH='x86_64-linux-gnu' FCFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong' LD_PRELOAD='' LANG='C' OBJCXXFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong -Wformat -Werror=format-security' DEB_TARGET_ARCH_ABI='base' LS_COLORS='' DEB_TARGET_ARCH_LIBC='gnu' PBUILDER_PKGLIBDIR='/usr/lib/pbuilder' SUDO_COMMAND='/usr/sbin/cowbuilder --build --buildresult /home/km/debian --debbuildopts  --debbuildopts  --basepath /var/cache/pbuilder/base.cow ../zimlib_3.3.0-1.dsc' DEB_BUILD_OPTIONS='parallel=4' DEBIAN_FRONTEND='noninteractive' DH_INTERNAL_BUILDFLAGS='1' CPPFLAGS='-Wdate-time -D_FORTIFY_SOURCE=2' SHELL='/bin/bash' GBP_BUILD_DIR='/home/km/debian/zimlib' DEB_HOST_ARCH_BITS='64' SUDO_USER='km' DEB_BUILD_ARCH='amd64' DEB_BUILD_GNU_CPU='x86_64' GBP_PBUILDER_DIST='sid' DEB_HOST_GNU_TYPE='x86_64-linux-gnu' COWDANCER_REUSE='yes' DEB_HOST_ARCH_CPU='amd64' DEB_RULES_REQUIRES_ROOT='binary-targets' PWD='/build/zimlib-3.3.0/obj-x86_64-linux-gnu' LC_ALL='C.UTF-8' GPG_TTY='/dev/pts/0' DEB_BUILD_MULTIARCH='x86_64-linux-gnu' FFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong' MFLAGS='' GCJFLAGS='-g -O2 -fdebug-prefix-map=/build/zimlib-3.3.0=. -fstack-protector-strong' DEB_HOST_ARCH_ENDIAN='little' PBCURRENTCOMMANDLINEOPERATION='build' DEB_TARGET_ARCH='amd64' /build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster
--- stdout ---
[==========] Running 7 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 5 tests from ClusterTest
[ RUN      ] ClusterTest.create_cluster
[       OK ] ClusterTest.create_cluster (0 ms)
[ RUN      ] ClusterTest.read_write_cluster
[       OK ] ClusterTest.read_write_cluster (0 ms)
[ RUN      ] ClusterTest.read_write_empty
[       OK ] ClusterTest.read_write_empty (0 ms)
[ RUN      ] ClusterTest.read_write_clusterZ
[       OK ] ClusterTest.read_write_clusterZ (1 ms)
[ RUN      ] ClusterTest.read_write_clusterLzma
[       OK ] ClusterTest.read_write_clusterLzma (22 ms)
[----------] 5 tests from ClusterTest (23 ms total)

[----------] 2 tests from CluterTest
[ RUN      ] CluterTest.read_write_extended_cluster
--- stderr ---

Assertion failed at ../src/cluster.cpp:85
 offset[4294968442] == reader->size().v[4294967295]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(_Z15_on_assert_failImmEvPKcS1_S1_T_T0_S1_i+0x15c) [0x564be5499e2c]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/../src/libzim.so.3(+0x1ed7e) [0x7f73b1c10d7e]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/../src/libzim.so.3(_ZN3zim7ClusterC2ESt10shared_ptrIKNS_6ReaderEENS_15CompressionTypeEb+0x9b) [0x7f73b1c0facb]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x1226e) [0x564be549626e]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x3c6ca) [0x564be54c06ca]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x32604) [0x564be54b6604]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x328c5) [0x564be54b68c5]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x32a70) [0x564be54b6a70]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x33075) [0x564be54b7075]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x3cb8a) [0x564be54c0b8a]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0x33296) [0x564be54b7296]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0xa13f) [0x564be548e13f]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f73b10a2a87]
/build/zimlib-3.3.0/obj-x86_64-linux-gnu/test/cluster(+0xa48a) [0x564be548e48a]
-------
```

**Comment by legoktm** [comment]
Oh, I can reproduce this on master as well.

**Comment by legoktm** [comment]
OK, after reading through the test case, I freed up all the possible memory on my computer (closed all of my open applications except a terminal), and it passed! I suspect the test is failing because I just don't have enough memory?? My computer usually has 10GB free, but I saw the test going up to 15GB (max of what I have).

Is there a way to make the tests significantly less memory intensive, or maybe a way to skip them?
